runtime/sema/check_variable_declaration.go

```diff
@@ -240,3 +242,155 @@ func (checker *Checker) elaborateNestedResourceMoveExpression(expression ast.Exp
+		//     ref2.field = 3
+		//
+		// Here, `ref2` refers to `ref1`, which refers to `v`.
+		// So `ref2` is actually referring to `v`
+
+		referencedVars := nestedReferencedVariables(referencedVariable)
+		if referencedVars != nil {
+			variables = append(variables, referencedVars...)
+		}
+	}
+
+	return
+}
+
+// referenceExpressions returns all sub-expressions that may produce a reference.
+//
+// There could be two types of expressions that can result in a reference:
+//  1. Expressions that create a new reference.
+//     (i.e: reference-expression)
+//  2. Expressions that return an existing reference.
+//     (i.e: identifier-expression/member-expression/index-expression having a reference type)
+//
+// However, it is currently not possible to track member-expressions and index-expressions.
+// So this method either returns a reference-expression or an identifier-expression.
+//
+// The expression could also be hidden inside some other expression.
+// e.g(1): `&v as &T` is a casting expression, but has a hidden reference expression.
+// e.g(2): `(&v as &T?)!
+func referenceExpressions(expr ast.Expression) []ast.Expression {
+	switch expr := expr.(type) {
+	case *ast.ReferenceExpression:
+		return []ast.Expression{expr}
+	case *ast.ForceExpression:
+		return referenceExpressions(expr.Expression)
+	case *ast.CastingExpression:
+		return referenceExpressions(expr.Expression)
+	case *ast.BinaryExpression:
+		if expr.Operation != ast.OperationNilCoalesce {
+			return nil
+		}
+
+		refExpressions := make([]ast.Expression, 0)
+
+		lhsRef := referenceExpressions(expr.Left)
+		if lhsRef != nil {
+			refExpressions = append(refExpressions, lhsRef...)
+		}
+
+		rhsRef := referenceExpressions(expr.Right)
+		if rhsRef != nil {
+			refExpressions = append(refExpressions, rhsRef...)
+		}
+
+		return refExpressions
+	case *ast.IdentifierExpression:
+		return []ast.Expression{expr}
+	default:
+		return nil
+	}
+}
+
+// rootVariableOfExpression returns the identifier expression
+// of a var-ref/member-access/index-access expression.
+func rootVariableOfExpression(expr ast.Expression) *ast.Identifier {
+	switch expr := expr.(type) {
+	case *ast.IdentifierExpression:
+		return &expr.Identifier
+	case *ast.MemberExpression:
+		return rootVariableOfExpression(expr.Expression)
+	case *ast.IndexExpression:
+		return rootVariableOfExpression(expr.TargetExpression)
+	default:
+		return nil
+	}
+}
+
+func nestedReferencedVariables(variable *Variable) []*Variable {
+	// If there are no more referenced variables, then it is the root of the reference chain.
+	if len(variable.referencedResourceVariables) == 0 {
+		// Add it as the referenced variable, if it is a resource.
+		if !variable.Type.IsResourceType() {
+			return nil
+		}
+
+		return []*Variable{variable}
+	}
+
+	var referencedResourceVariables []*Variable
+	for _, referencedVar := range variable.referencedResourceVariables {
+		nestedReferencedVars := nestedReferencedVariables(referencedVar)
```

Comment:
the reference chain is a tree, and needs to visit in a depth-first manner. Writing this with a for loop would not be more optimized than the recursion, as it would need to maintain a stack.

Recursion here might not be that bad, because in practice, it is highly unlikely people would write code that has long chains of references. Maybe a depth of one or two in most cases.

Reply:
Ah, right. I missed that this was tree-shaped 👍